grails-app/controllers/org/pih/warehouse/invoice/InvoiceController.groovy

```diff
@@ -64,6 +67,11 @@ class InvoiceController {
         def invoiceInstance = Invoice.get(params.id)
         if (invoiceInstance) {
             try {
+                invoiceInstance.invoiceItems?.each { InvoiceItem invoiceItem ->
```

Comment:
Seeing this again really worries me. For now can you at least move lines 70-75 to the InvoiceService and just call invoiceService.deleteInvoice(invoice) here.

Reply:
With the new deleteInvoiceItem(InvoiceItem invoiceItem) method suggested below ... this new method InvoiceService.deleteInvoice(Invoice) should look like this
```
void deleteInvoice(Invoice invoice) {
    invoice.invoiceItems.each { InvoiceItem invoiceItem ->
        deleteInvoiceItem(invoiceItem)
    }
}
```

Reply:
done